zetaclient/orchestrator/bootstrap.go

```diff
@@ -0,0 +1,414 @@
+package orchestrator
+
+import (
+	"context"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	solrpc "github.com/gagliardetto/solana-go/rpc"
+	"github.com/pkg/errors"
+
+	"github.com/zeta-chain/zetacore/zetaclient/chains/base"
+	btcobserver "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/observer"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/rpc"
+	btcsigner "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/signer"
+	evmobserver "github.com/zeta-chain/zetacore/zetaclient/chains/evm/observer"
+	evmsigner "github.com/zeta-chain/zetacore/zetaclient/chains/evm/signer"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/interfaces"
+	solbserver "github.com/zeta-chain/zetacore/zetaclient/chains/solana/observer"
+	"github.com/zeta-chain/zetacore/zetaclient/config"
+	zctx "github.com/zeta-chain/zetacore/zetaclient/context"
+	"github.com/zeta-chain/zetacore/zetaclient/db"
+	"github.com/zeta-chain/zetacore/zetaclient/metrics"
+)
+
+// backwards compatibility
+const btcDatabaseFilename = "btc_chain_client"
+
+// CreateSignerMap creates a map of interfaces.ChainSigner (by chainID) for all chains in the config.
+// Note that signer construction failure for a chain does not prevent the creation of signers for other chains.
+func CreateSignerMap(
+	ctx context.Context,
+	tss interfaces.TSSSigner,
+	logger base.Logger,
+	ts *metrics.TelemetryServer,
+) (map[int64]interfaces.ChainSigner, error) {
+	signers := make(map[int64]interfaces.ChainSigner)
+	_, _, err := syncSignerMap(ctx, tss, logger, ts, &signers)
+	if err != nil {
+		return nil, err
+	}
+
+	return signers, nil
+}
+
+// syncSignerMap synchronizes the given signers map with the signers for all chains in the config.
+// This semantic is used to allow dynamic updates to the signers map.
+// Note that data race handling is the responsibility of the caller.
+func syncSignerMap(
+	ctx context.Context,
+	tss interfaces.TSSSigner,
+	logger base.Logger,
+	ts *metrics.TelemetryServer,
+	signers *map[int64]interfaces.ChainSigner,
+) (int, int, error) {
+	if signers == nil {
+		return 0, 0, errors.New("signers map is nil")
+	}
+
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return 0, 0, errors.Wrapf(err, "failed to get app context")
+	}
+
+	var (
+		added, removed int
+
+		presentChainIDs = make([]int64, 0)
+
+		onAfterAdd = func(chainID int64, _ interfaces.ChainSigner) {
+			logger.Std.Info().Msgf("Added signer for chain %d", chainID)
+			added++
+		}
+
+		addSigner = func(chainID int64, signer interfaces.ChainSigner) {
+			mapSet[int64, interfaces.ChainSigner](signers, chainID, signer, onAfterAdd)
+		}
+
+		onBeforeRemove = func(chainID int64, _ interfaces.ChainSigner) {
+			logger.Std.Info().Msgf("Removing signer for chain %d", chainID)
+			removed++
+		}
+	)
+
+	// EVM signers
+	for _, evmConfig := range app.Config().GetAllEVMConfigs() {
```

Comment:
Just to track. The config file reloader would be a piece of future refactor to allow the observer signer to feed config (e.g. their own endpoint for the chain) of newly enabled chains.

Reply:
https://github.com/zeta-chain/node/issues/2519